src/integration-test/java/com/commercetools/sync/integration/commons/utils/ProductITUtils.java

```diff
@@ -48,19 +48,22 @@ public static void deleteProductSyncTestData(@Nonnull final SphereClient ctpClie
     }
 
     /**
-     * Deletes all products from the CTP project defined by the {@code ctpClient}.
+     * Unpublishes all published products, then deletes all products from the CTP project defined by the
+     * {@code ctpClient}.
      *
-     * @param ctpClient defines the CTP project to delete the categories from.
+     * @param ctpClient defines the CTP project to delete the products from.
      */
     public static void deleteAllProducts(@Nonnull final SphereClient ctpClient) {
-        final Function<Product, SphereRequest<Product>> unpublishAndDelete = product -> {
-            if (product.getMasterData().isPublished()) {
-                product = ctpClient.execute(ProductUpdateCommand.of(product, Unpublish.of()))
-                                   .toCompletableFuture().join();
-            }
-            return ProductDeleteCommand.of(product);
-        };
-        queryAndApply(ctpClient, ProductQuery::of, unpublishAndDelete);
+        unPublishAllPublishedProducts(ctpClient);
+        queryAndApply(ctpClient, ProductQuery::of, ProductDeleteCommand::of);
```

Comment:
You can not delete unpublished product. If you do it in parallel then there is no guarantee that on each delete product has been published already. You can create a future that un-publishes und deletes single product and internally chains successful un-publish to deletion command. This method can be then called in parallel for different products that were returned by publishedProductsQuery command.

Reply:
the [`queryAndApply`](https://github.com/commercetools/commercetools-sync-java/blob/master/src/integration-test/java/com/commercetools/sync/integration/commons/utils/ITUtils.java#L68) joins the completebleFuture. So all products are unpublished all in parallel before being (deleted all in parallel).